HEMS: add templates

andig opened the pull request:
Adds HEMS configuration via the UI: pick a HEMS integration (EEBUS, HEMS Pro, GPIO relay, FNN steuerbox) or a user-defined YAML one. Legacy YAML configs migrate automatically.

- Template-based modal replaces the YAML-only one; edit icon next to the integration selector lets users swap providers
- Card shows the active integration name; External Limit moved next to Load Management
- Backend templates + migration of legacy keys.Hems settings to device entries
- Playwright coverage for the new flows

**Todos**

- [x] check added templates @andig @premultiply 
- [x] decide on hems device status, currently none (limit?) @andig see https://github.com/evcc-io/evcc/pull/30284
- [x] add a hint that curtailment limits **MUST** be configured per device atm, maybe link to docs @naltatis 

**Screenshots**

overview

<img width="973" height="407" alt="overview" src="https://github.com/user-attachments/assets/453d02e3-a734-4c6e-aa59-0866162ebe9f" />

template select

<img width="760" height="663" alt="select" src="https://github.com/user-attachments/assets/b7227c6d-a4ca-4f1a-80ff-b42e27d1978f" />

custom type

<img width="934" height="1187" alt="custom" src="https://github.com/user-attachments/assets/1de5babe-feda-405b-b9d4-1e44021b70e6" />

replace device

<img width="1004" height="530" alt="replace" src="https://github.com/user-attachments/assets/cea62c82-af2b-4da7-a8d9-eec5092d2c9d" />

template

<img width="648" height="1154" alt="fnn" src="https://github.com/user-attachments/assets/04dda692-e976-4d2b-8bf8-4e7526141bfd" />

andig committed:
HEMS: add templates

andig committed:
wip

andig committed:
wip

andig committed:
Add services

andig committed:
Merge branch 'master' into feat/hems-tmpl

andig committed:
wip

naltatis committed:
Merge branch 'master' into feat/hems-tmpl

naltatis committed:
fix e2e

naltatis committed:
fix e2e

andig commented:
ping @naltatis bit-rotten :(

naltatis committed:
Merge branch 'master' into feat/hems-tmpl

naltatis committed:
Merge branch 'master' into feat/hems-tmpl

naltatis committed:
add ui and tests

naltatis committed:
handle fatals; show integration name in card

naltatis committed:
lint

naltatis committed:
ping

naltatis committed:
review feedback

naltatis commented on cmd/setup.go:
```diff
@@ -775,36 +775,83 @@ func configureGo(conf []globalconfig.Go) error {
 	}
 
-	typ, other, err := config.CustomDevice(conf.Type, conf.Other)
+	return configureDbHems(conf, site)
```
YamlSourceDb is not applicable here. we are moving from yaml to devices. source can be removed in future

naltatis commented:
Updated PR description and added screenshots. @andig I had to make quite a few go changes to make CURD + test work reliably. Please check. UI-side, checks, migration should all be fine now.

naltatis commented:
@copilot resolve merge conflict

Copilot committed:
Merge origin/master and resolve conflicts

naltatis committed:
Merge branch 'master' into feat/hems-tmpl

andig committed:
Update eebus.go

andig commented on cmd/setup.go:
```diff
@@ -775,36 +775,83 @@ func configureGo(conf []globalconfig.Go) error {
+		return nil, err
+	}
+
+	dev := configurable[0]
```
Das sieht alles sehr merkwürdig aus- solche Patterns haben wir sonst auch nicht- warum brauchts das?

naltatis replied:
Ja, wir haben hier den neuen Fall, dass wir Geräte erzeugen die nirgendwo angemeldet/referenziert sind und dass es immer nur genau null oder eins geben kann. Alternativ können wir das auch in `config` verschieben und da singular Methoden anbieten: `config.ConfigurationByClass(templates.Hems)`.

andig replied:
Unsure. Wenns mehrere gibt- wieso ist das Nullte richtig?

naltatis replied:
Habs jetzt mal ausgelagert. Die Mechanik bleibt aber.

naltatis replied:
> Unsure. Wenns mehrere gibt- wieso ist das Nullte richtig?

Wenn wir mehrere in der DB haben haben wir eine Konsistenzproblem.

andig replied:
Is klar- aber wieso ist das Nullte das welches wir wollen? Dann evtl. lieber alle (uninitialisiert) rot anzeigen und löschen lassen?

naltatis replied:
Alternative Vorschläge? Optionen sind "first", "last" oder "random" 🤣

naltatis replied:
> Dann evtl. lieber alle (uninitialisiert) rot anzeigen und löschen lassen?

Wir wollen dem Nutzer doch keine UI für Fehlkonfigurationsmöglichkeiten unseres generischen Datenmodells geben. Die Konsistenz müssen wir beim Eingang (Geräte anlegen) sicherstellen. Auf UI Ebene ist das bereits gegeben. Auf API Ebene kann ich das noch einmal prüfen.

naltatis replied:
> Auf API Ebene kann ich das noch einmal prüfen.

Im POST API (create device) habe ich jetzt einen Check hinzugefügt der das auch bei direkter API Nutzung sicherstellt.

andig commented on cmd/setup.go:
```diff
@@ -775,36 +775,83 @@ func configureGo(conf []globalconfig.Go) error {
+	typ, other, err := config.CustomDevice(cc.Type, cc.Other)
 	if err != nil {
-		return nil, fmt.Errorf("failed configuring hems: %w", err)
+		err = fmt.Errorf("hems '%s': %w", cc.Name, err)
```
Fehlermeldung weicht von anderen Gerätetypen ab- warum?

andig commented on cmd/setup.go:
```diff
@@ -775,36 +775,83 @@ func configureGo(conf []globalconfig.Go) error {
+		return nil, err
+	}
+
+	conf.Type, conf.Other = typ, other
```
Why?

naltatis replied:
unnecessary. also eliminates the need of conf here over all. deleted

andig commented on hems/fnn/fnn.go:
```diff
@@ -9,13 +9,20 @@ import (
+
 // NewFromConfig creates an FNN HEMS from generic config.
-func NewFromConfig(ctx context.Context, other map[string]any, site site.API) (*Fnn, error) {
+func NewFromConfig(ctx context.Context, other map[string]any, site site.API) (hems.API, error) {
```
Why change return type? Go patten ist "return concrete types"- if you can.

naltatis replied:
Result has to go into `config.Registry.AddCtx` where hems.API type is needed. Maybe I'm missing something but this is the same way we do it for other devices - or not?

```
func NewKebaFromConfig(ctx context.Context, other map[string]any) (api.Charger, error) {
```

andig replied:
Needed this due to decoration. For HEMS nothing is decorated. Always return concrete type, not interface if possible. We could update all chargers/meters too, if we wanted...

naltatis replied:
Ok, it's unclear for me what we're gaining here. I've now changed the types and added a wrapper for type conversion (registry).

naltatis committed:
review feedback: cleanup

naltatis committed:
review: better errors

naltatis committed:
extract ConfigurationByClass

naltatis committed:
concrete types

naltatis committed:
api: ensure hems singleton

andig commented:
@copilot resolve the merge conflicts in this pull request

Copilot committed:
Merge origin/master and resolve Config.vue conflict

andig committed:
Merge origin/master into feat/hems-tmpl

Resolve conflicts after the hems/circuit split (#30284) removed the
gridcontrol circuit: wire HEMS via site.SetHEMS instead of
smartgrid.SetupCircuit, keep the template-based config (Ip field,
config.AddCtx, db hems device), and republish runtime dim/curtail
status. Adapt the fnn/curtailment e2e tests to the new HEMS modal flow.

andig committed:
HEMS: align fnn-3 template with fnn implementation

Render the current maxcurtailpower field instead of the deprecated
maxpower, and emit type: fnn (fnn-3 is registered as an alias).

andig committed:
HEMS: remove hemspro template

andig commented on templates/definition/hems/fnn.yaml:
```diff
@@ -0,0 +1,59 @@
+      en: Maximum curtailment power
+      de: Maximale Reduzierungsleistung
+    help:
+      en: Full production power. Curtailment steps (60%, 30%, 0%) apply to this baseline.
```
Full seems wrong- it's the maximum allowed feedin power. @premultiply ?

naltatis replied:
@premultiply ping

andig replied:
> Reduzierungsleistung

Ich bin nicht sicher wie die richtigen Begriffe lauten, aber ich hätte keien Ahnung was das sein soll.

andig commented on templates/definition/hems/fnn.yaml:
```diff
@@ -0,0 +1,59 @@
+    unit: W
+    required: true
+    description:
+      en: Maximum curtailment power
```
Rather "curtailment power limit"?

andig commented:
@naltatis tried to resolve the merge conflicts. Please check setup.go.

CiNcH83 commented:
>External Limit moved next to Load Management

[EDIT] Sorry. Have been on the wrong branch. Order is correct.

CiNcH83 commented:
I tried to add controlbox with the new UI, but failed:
<img width="468" height="832" alt="image" src="https://github.com/user-attachments/assets/cbb36aa9-a308-4e5d-a6a7-2528d6efb533" />

It is getting stuck. Do I have to unpair the previous pairing first?

[trace_log.log](https://github.com/user-attachments/files/28788390/trace_log.log)

andig commented on templates/definition/hems/fnn-3.yaml:
```diff
@@ -0,0 +1,59 @@
+template: fnn-3
```
Should be FNN now

andig commented on templates/definition/hems/relay.yaml:
```diff
@@ -0,0 +1,46 @@
+      en: Maximum power
+      de: Maximale Leistung
+    help:
+      en: Power cap applied while the relay signal is active
```
applied where? the root circuit.

andig commented:
No, you don't need to unpair anything first — the previous pairing isn't the problem here.

Your trace log shows the actual cause at startup:

```
[eebus] ERROR websocket server error: listen tcp :4712: bind: address already in use
```

evcc's EEBus SHIP server defaults to port **4712**, but that port is already taken — by the ControlBox **Simulator** itself, which is running on the same machine and advertising on 4712:

```
mdns: Demo-ControlBox-123456789 ... 127.0.0.1 4712 ... model:ControlBox type:ElectricitySupplySystem
```

(The `172.17/18/19.0.1` addresses for it show it's a Docker container on the same host.) Both evcc and the Demo ControlBox simulator default to 4712, so they collide. With its server unable to bind, evcc has no SHIP endpoint to complete the handshake, so pairing just hangs.

Two ways out:

1. Give the simulator a different host port (e.g. publish the container as `-p 4713:4712`) so 4712 is free for evcc, or run it on another machine. Real ControlBox hardware lives on its own device, so this only bites when the simulator shares the host with evcc.
2. Or move evcc's EEBus port via the global `eebus:` config (`port:`), then retry discovery.

Separately, evcc should make this obvious instead of only logging it — I opened #30838 so the EEBus server surfaces a proper error state when it can't start.

andig committed:
Merge branch 'master' into feat/hems-tmpl

CiNcH83 commented:
Ah OK. Yes. I run controlbox and evcc test branches on the same Raspberry Pi...

naltatis commented:
> @naltatis tried to resolve the merge conflicts. Please check setup.go.

Looks plausible - and all tests pass.

naltatis committed:
review feedback

naltatis committed:
wording

naltatis commented:
@copilot resolve the merge conflicts in this pull request

Copilot committed:
Merge master into feat/hems-tmpl

premultiply committed:
wip: templates

premultiply committed:
wip: (reunify) fnn templates

premultiply committed:
wip: drop legacy relay template

premultiply committed:
fix: eebus failsafe-default value

premultiply committed:
wip: improve fnn-eebus template

premultiply committed:
Merge remote-tracking branch 'origin/master' into feat/hems-tmpl

# Conflicts:
#	assets/js/components/Config/DeviceModal/DeviceModalBase.vue
#	hems/eebus/eebus.go
#	server/http.go
#	server/http_config_device_handler.go

premultiply committed:
wip: add hemspro templates

premultiply committed:
fix eebus test

premultiply reviewed (commented):
Ich bin erstmal durch mit den Änderungsvorschlägen

naltatis committed:
Merge branch 'master' into feat/hems-tmpl

naltatis commented:
Once again checked the code to verify that this is not breaking (migrated) for existing configurations. Looks fine.
